feat: python udf support

joker-star-l opened the pull request:
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #9601 
issue #11900

## What this PR does / why we need it:

1. feat: python udf 实现

* 创建函数
```
-- 内嵌式，function body 直接写在 sql 语句中
create or replace function py_add (a int, b int) returns int language python as 
$$
def add(a, b):
  return a + b
$$
handler 'add';

-- 导入式，从本地导入一个 *.py 文件或者 *.whl 包
create or replace function py_add (a int, b int) returns int language python import 
'./add.py' 
handler 'test';  -- 调用 test.py 中的 add 函数

create or replace function py_add (a int, b int) returns int language python import 
'./test-1.0.0-py3-none-any.whl' 
handler 'test.add';  -- 调用 test-1.0.0-py3-none-any.whl 包 test 模块/文件中的 add 函数
```

* 调用函数
```
select py_add(1, 1);
-- 输出：
+--------------+
| py_add(1, 1) |
+--------------+
|            2 |
+--------------+
```

* vector api
```
-- 在创建函数时使用 function_name.vector=True 来标记，表示该函数接收的参数为 vector 形式（python list）
create or replace function py_vecadd(a int, b int) returns int language python as 
$$
def add(a, b):
    ret = [a[i] + b[i] for i in range(len(a))]
    return ret
add.vector = True
$$
handler 'add';
```

2. refactor: 表 `mo_catalog.mo_user_defined_function` 字段 `args` 优化（break change）
    1. 类型由 `text` 修改为 `json`
    2. 格式从 `{"a":"int","b":"int"}` 修改为 `[{"name":"a","type":"int"},{"name":"b","type":"int"}]`（以 `py_add(a int, b int)` 为例）。原因是前者无法保存参数顺序，在进行函数匹配时会出现 bug
 
3. fix: 在执行内部 sql 时对 json 字段进行深拷贝

joker-star-l commented:
@fengttt @aylei @iamlinjunhong @aunjgr

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/matrixorigin/matrixone?pullRequest=11098) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/matrixorigin/matrixone?pullRequest=11098) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/matrixorigin/matrixone?pullRequest=11098) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/matrixorigin/matrixone?pullRequest=11098) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/matrixorigin/matrixone?pullRequest=11098) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/matrixorigin/matrixone?pullRequest=11098) it.</sub>

fengttt reviewed (commented):
When/Where do we start the python server?

joker-star-l commented:
> When/Where do we start the python server?

每个 CN 配一个 python server，作为 CN 的 side car 与 CN 同时启动

aylei commented on pkg/pythonserver/server.py:
```diff
@@ -0,0 +1,121 @@
+
+import grpc
+
+import python_udf_pb2_grpc as pb2_grpc
```
why pb2?

joker-star-l replied:
this name is generated by grpc plugin

aylei commented on pkg/pythonserver/server.py:
```diff
@@ -0,0 +1,121 @@
+
+
+if __name__ == '__main__':
+    run()
```
when do we start this python server

joker-star-l replied:
at the same time with cn starting

fengttt reviewed (commented):
Where are tests?

Have you been able to setup the python side car thing for to run tests?

joker-star-l commented:
> Where are tests?
> 
> Have you been able to setup the python side car thing for to run tests?

Currently, I just setup python server manually and test some function examples in local environment. I will do some work on how to automatically setup python server without k8s and write some test code in this week.

joker-star-l commented:
In commit joker-star-l@077bfa2, I added a "UdfService" to handle non-sql udf in cn (currently just python udf) and put interface `UdfService` into struct `Process`. 
However, in order to do this, **I added the arg `UdfService` in many functions**. 
@aylei and I have had a discussion but we are not sure whether this change is reasonable, could you help me take a look? Thanks! @iamlinjunhong

aylei commented on pkg/udf/service.go:
```diff
@@ -0,0 +1,46 @@
+
+// Service handle non-sql udf in cn
+type Service interface {
+	AddPythonUdfClient(pc pythonservice.PythonUdfClient)
```
This method leak implementation details to the caller of UDF interface, I'd suggest remove this and manage the client in the implementation of this interface.

joker-star-l replied:
Here `PythonUdfClient` is just a interface.

aylei replied:
> Here `PythonUdfClient` is just a interface.

why another interface?

aylei commented on pkg/udf/service.go:
```diff
@@ -0,0 +1,46 @@
+// Service handle non-sql udf in cn
+type Service interface {
+	AddPythonUdfClient(pc pythonservice.PythonUdfClient)
+	RunPythonUdf(ctx context.Context, request *pythonservice.PythonUdfRequest) (*pythonservice.PythonUdfResponse, error)
```
Since we are building an UDF interface, I think `RunUdf()` would be better, so does the parameter type and return type.

aylei reviewed (commented):
The rest LGTM, could you please add UT cases to pass the UT coverage check?

aylei commented on pkg/udf/service.go:
```diff
@@ -0,0 +1,46 @@
+)
+
+// Service handle non-sql udf in cn
+type Service interface {
```
@iamlinjunhong please take a look in favor of future UDAF and UDTF support, thanks!

iamlinjunhong replied:
get it

iamlinjunhong commented:
LGTM, but BVT and UT

joker-star-l commented:
> The rest LGTM, could you please add UT cases to pass the UT coverage check?

I am doing this currently, I just find out that the coverage is the sum of BVT and UT.  I have write some code to run BVT locally and calculate the coverage of UT and BVT together, which would help me to find out where I need to add test case.

joker-star-l commented:
> LGTM, but BVT and UT

Ok, thanks!

aylei reviewed (approved):
lgtm, awesome work

daviszhen commented on pkg/frontend/mysql_cmd_executor.go:
```diff
@@ -1290,15 +1291,70 @@ func (mce *MysqlCmdExecutor) handleDropRole(ctx context.Context, dr *tree.DropRo
 
-	return InitFunction(ctx, ses, tenant, cf)
+	return InitFunction(ctx, ses, tenant, cf, func(localPath string, storageDir string) (string, error) {
+		loadLocalReader, loadLocalWriter := io.Pipe()
```
这个嵌入的函数代码有点长。独立成单独的函数吧。

joker-star-l replied:
好的

daviszhen replied:
upload 独立成单独的函数。不要放到这个函数里面。太长了。

joker-star-l replied:
不太行，放到这个函数里是因为需要外面函数的上下文。。。

sukki37 commented on etc/launch-with-python-udf-server/cn.toml:
```diff
@@ -0,0 +1,12 @@
+port-base = 18000
+
+[cn.python-udf-client]
+server-address = "127.0.0.1:50051"
```
Please consider using auto-generated port based on 'port-base' for internal communication ports

joker-star-l replied:
This config is only for client. It is independent with the python server config. So should I change the both to 'port-base' or only change the server config (in file 'python-udf-server.toml')?

aylei replied:
Python UDF server address is configured outside the CN process and will not interact with HAKeeper (since it is design to have coupled lifecycle with CN), so the server-address configuration is required if python UDF server is not running on the default address (if any)

daviszhen commented on pkg/frontend/authenticate.go:
```diff
@@ -8328,14 +8370,63 @@ func InitRole(ctx context.Context, ses *Session, tenant *TenantInfo, cr *tree.Cr
+	// watch and cancel
+	// TODO use context.AfterFunc in go1.21
+	go func() {
+		defer loadLocalReader.Close()
```
正常退出的时候，routine没有在函数退出前清理掉。

daviszhen commented on pkg/frontend/authenticate.go:
```diff
@@ -8328,14 +8370,63 @@ func InitRole(ctx context.Context, ses *Session, tenant *TenantInfo, cr *tree.Cr
+	return InitFunction(ctx, ses, tenant, cf, mce)
+}
+
+func InitFunction(ctx context.Context, ses *Session, tenant *TenantInfo, cf *tree.CreateFunction, uploader udf.PkgUploader) (err error) {
```
InitFunction 和 func (mce *MysqlCmdExecutor) InitFunction 合并掉。

EricDrui commented:
When I was using the *.whl package, I successfully created a function but the call prompt module did not exist. What is the reason for this？
![1703494449069](https://github.com/matrixorigin/matrixone/assets/29595389/a7a828ad-0c32-452d-ad93-83cf4f3ec761)
![1703494478749](https://github.com/matrixorigin/matrixone/assets/29595389/5f18666d-52ec-429b-b4c2-77fce3aef261)

joker-star-l commented:
> When I was using the *.whl package, I successfully created a function but the call prompt module did not exist. What is the reason for this？ ![1703494449069](https://private-user-images.githubusercontent.com/29595389/292737845-a7a828ad-0c32-452d-ad93-83cf4f3ec761.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM0OTc0NzcsIm5iZiI6MTcwMzQ5NzE3NywicGF0aCI6Ii8yOTU5NTM4OS8yOTI3Mzc4NDUtYTdhODI4YWQtMGMzMi00NTJkLWFkOTMtODNjZjRmM2VjNzYxLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI1VDA5MzkzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzZjExOWJhZDQ3M2I1NjZhYTA3YWZjNjcxNDZmZmM1OTVjNjlkOTAyNjc3MmRlNjE1YTdlZDkxZDc5YmRkY2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._abGJkJf9J5rU1bIjN-cmv-tEbc7Krp_IR4l1jSn-i8) ![1703494478749](https://private-user-images.githubusercontent.com/29595389/292737848-5f18666d-52ec-429b-b4c2-77fce3aef261.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM0OTc0NzcsIm5iZiI6MTcwMzQ5NzE3NywicGF0aCI6Ii8yOTU5NTM4OS8yOTI3Mzc4NDgtNWYxODY2NmQtNTJlYy00MjliLWI0YzItNzdmY2UzYWVmMjYxLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI1VDA5MzkzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQwMGNmMmMyZTM2YjM2YzEwNjJlZWJhNTIxMTQ3YWNlNTA4NTc4ODk3YmZkYTA5ZmNkZDE1MTg3M2I0MWZlYzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.on5A5XSLpscWF0p-sp_eD-g4LBoanXUjDz_ENgVa6r8) [mathlib-1.0.0-py3-none-any.zip](https://github.com/matrixorigin/matrixone/files/13765259/mathlib-1.0.0-py3-none-any.zip)

Hello, maybe in the `__init__.py`, it should be `from .mathlib ...`, which means a relative path.

EricDrui commented:
> > When I was using the *.whl package, I successfully created a function but the call prompt module did not exist. What is the reason for this？ ![1703494449069](https://private-user-images.githubusercontent.com/29595389/292737845-a7a828ad-0c32-452d-ad93-83cf4f3ec761.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM0OTc0NzcsIm5iZiI6MTcwMzQ5NzE3NywicGF0aCI6Ii8yOTU5NTM4OS8yOTI3Mzc4NDUtYTdhODI4YWQtMGMzMi00NTJkLWFkOTMtODNjZjRmM2VjNzYxLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI1VDA5MzkzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzZjExOWJhZDQ3M2I1NjZhYTA3YWZjNjcxNDZmZmM1OTVjNjlkOTAyNjc3MmRlNjE1YTdlZDkxZDc5YmRkY2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._abGJkJf9J5rU1bIjN-cmv-tEbc7Krp_IR4l1jSn-i8) ![1703494478749](https://private-user-images.githubusercontent.com/29595389/292737848-5f18666d-52ec-429b-b4c2-77fce3aef261.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM0OTc0NzcsIm5iZiI6MTcwMzQ5NzE3NywicGF0aCI6Ii8yOTU5NTM4OS8yOTI3Mzc4NDgtNWYxODY2NmQtNTJlYy00MjliLWI0YzItNzdmY2UzYWVmMjYxLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI1VDA5MzkzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQwMGNmMmMyZTM2YjM2YzEwNjJlZWJhNTIxMTQ3YWNlNTA4NTc4ODk3YmZkYTA5ZmNkZDE1MTg3M2I0MWZlYzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.on5A5XSLpscWF0p-sp_eD-g4LBoanXUjDz_ENgVa6r8) [mathlib-1.0.0-py3-none-any.zip](https://github.com/matrixorigin/matrixone/files/13765259/mathlib-1.0.0-py3-none-any.zip)
> 
> Hello, maybe in the `__init__.py`, it should be `from .mathlib ...`, which means a relative path.

Thank you for your reply. Following your instructions, it can now run normally. Thanks!
